api-gateway-config/scripts/lua/management/subscriptions.lua

```diff
@@ -0,0 +1,122 @@
+--   all copies or substantial portions of the Software.
+--
+--   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+--   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+--   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+--   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+--   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+--   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+--   DEALINGS IN THE SOFTWARE.
+
+--- @module subscriptions
+-- Management interface for subscriptions for the gateway
+
+local cjson = require "cjson"
+local redis = require "lib/redis"
+local utils = require "lib/utils"
+local request = require "lib/request"
+
+local REDIS_HOST = os.getenv("REDIS_HOST")
+local REDIS_PORT = os.getenv("REDIS_PORT")
+local REDIS_PASS = os.getenv("REDIS_PASS")
+
+local _M = {}
+
+--- Add an apikey/subscription to redis
+-- PUT /v1/subscriptions
+-- Body:
+-- {
+--    key: *(String) key for tenant/api/resource
+--    scope: *(String) tenant or api or resource
+--    tenantId: *(String) tenant id
+--    resource: (String) url-encoded resource path
+--    apiId: (String) api id
+-- }
+function _M.addSubscription()
+  -- Validate body and create redisKey
+  local redisKey = validateSubscriptionBody()
+  -- Open connection to redis or use one from connection pool
+  local red = redis.init(REDIS_HOST, REDIS_PORT, REDIS_PASS, 10000)
+  redis.createSubscription(red, redisKey)
+  -- Add current redis connection in the ngx_lua cosocket connection pool
+  redis.close(red)
+  request.success(200, "Subscription created.")
+end
+
+--- Delete apikey/subscription from redis
+-- DELETE /v1/subscriptions
+-- Body:
+-- {
+--    key: *(String) key for tenant/api/resource
+--    scope: *(String) tenant or api or resource
+--    tenantId: *(String) tenant id
+--    resource: (String) url-encoded resource path
+--    apiId: (String) api id
+-- }
+function _M.deleteSubscription()
+  -- Validate body and create redisKey
+  local redisKey = validateSubscriptionBody()
+  -- Initialize and connect to redis
+  local red = redis.init(REDIS_HOST, REDIS_PORT, REDIS_PASS, 10000)
+  -- Return if subscription doesn't exist
+  redis.deleteSubscription(red, redisKey)
+  -- Add current redis connection in the ngx_lua cosocket connection pool
+  redis.close(red)
+  request.success(200, "Subscription deleted.")
+end
+
+--- Check the request JSON body for correct fields
+-- @return redisKey subscription key for redis
+function validateSubscriptionBody()
+  -- Read in the PUT JSON Body
+  ngx.req.read_body()
+  local args = ngx.req.get_body_data()
+  if not args then
+    request.err(400, "Missing request body.")
+  end
+  -- Convert json into Lua table
+  local decoded = cjson.decode(args)
+  -- Check required fields
+  local requiredFieldList = {"key", "scope", "tenantId"}
```

Comment:
I think we can extract the input validation to make it reusable. I'm seeing the same logic in other parts, i.e. https://github.com/openwhisk/apigateway/pull/62/files#diff-eca0bb209f1cd0d1f6c13768183f866eR62